File: deploy/ultra-infer/python/ultra_infer/vision/perception/paddle3d/centerpoint.py
# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import absolute_import

from .... import UltraInferModel, ModelFormat
from .... import c_lib_wrap as C


class CenterpointPreprocessor:
    def __init__(self, config_file):
        """Create a preprocessor for Centerpoint"""
        self._preprocessor = C.vision.perception.CenterpointPreprocessor(config_file)

    def run(self, point_dirs, num_point_dim, with_timelag):
        """Preprocess input images for Centerpoint

        :param: input_ims: (list of numpy.ndarray)The input image
        :return: list of FDTensor
        """
        return self._preprocessor.run(point_dirs, num_point_dim, with_timelag)


class Centerpoint(UltraInferModel):
    def __init__(
        self,
        model_file,
        params_file,
        config_file,
        runtime_option=None,
        model_format=ModelFormat.PADDLE,
    ):
        """Load a Centerpoint model exported by Centerpoint.

        :param model_file: (str)Path of model file, e.g ./Centerpoint.pdmodel
        :param params_file: (str)Path of parameters file, e.g ./Centerpoint.pdiparams
        :param config_file: (str)Path of config file, e.g ./infer_cfg.yaml
        :param runtime_option: (ultra_infer.RuntimeOption)RuntimeOption for inference this model, if it's None, will use the default backend on CPU
        :param model_format: (ultra_infer.ModelForamt)Model format of the loaded model
        """
        super(Centerpoint, self).__init__(runtime_option)

        self._model = C.vision.perception.Centerpoint(
            model_file, params_file, config_file, self._runtime_option, model_format
        )
        assert self.initialized, "Centerpoint initialize failed."

    def predict(self, point_dir):
        """Detect an input image

        :param input_image: (numpy.ndarray)The input image data, 3-D array with layout HWC, BGR format
        :param conf_threshold: confidence threshold for postprocessing, default is 0.25
        :param nms_iou_threshold: iou threshold for NMS, default is 0.5
        :return: PerceptionResult
        """
        return self._model.predict(point_dir)

    def batch_predict(self, points_dir):
        """Classify a batch of input image

        :param im: (list of numpy.ndarray) The input image list, each element is a 3-D array with layout HWC, BGR format
        :return list of PerceptionResult
        """

        return self._model.batch_predict(points_dir)

    @property
    def preprocessor(self):
        """Get CenterpointPreprocessor object of the loaded model

        :return CenterpointPreprocessor
        """
        return self._model.preprocessor

    @property
    def postprocessor(self):
        """Get CenterpointPostprocessor object of the loaded model

        :return CenterpointPostprocessor
        """
        return self._model.postprocessor

File: deploy/ultra-infer/ultra_infer/vision/perception/paddle3d/centerpoint/centerpoint.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ultra_infer/vision/perception/paddle3d/centerpoint/centerpoint.h"

namespace ultra_infer {
namespace vision {
namespace perception {

Centerpoint::Centerpoint(const std::string &model_file,
                         const std::string &params_file,
                         const std::string &config_file,
                         const RuntimeOption &custom_option,
                         const ModelFormat &model_format)
    : preprocessor_(config_file) {
  valid_gpu_backends = {Backend::PDINFER};

  runtime_option = custom_option;
  runtime_option.model_format = model_format;
  runtime_option.model_file = model_file;
  runtime_option.params_file = params_file;
  initialized = Initialize();
}

bool Centerpoint::Initialize() {
  if (!InitRuntime()) {
    FDERROR << "Failed to initialize ultra_infer backend." << std::endl;
    return false;
  }
  return true;
}

bool Centerpoint::Predict(const std::string point_dir,
                          PerceptionResult *result) {
  std::vector<PerceptionResult> results;
  if (!BatchPredict({point_dir}, &results)) {
    return false;
  }

  if (results.size()) {
    *result = std::move(results[0]);
  }
  return true;
}

bool Centerpoint::BatchPredict(std::vector<std::string> points_dir,
                               std::vector<PerceptionResult> *results) {
  int64_t num_point_dim = 5;
  int with_timelag = 0;
  if (!preprocessor_.Run(points_dir, num_point_dim, with_timelag,
                         reused_input_tensors_)) {
    FDERROR << "Failed to preprocess the input image." << std::endl;
    return false;
  }

  results->resize(reused_input_tensors_.size());
  for (int index = 0; index < reused_input_tensors_.size(); ++index) {
    std::vector<FDTensor> input_tensor;
    input_tensor.push_back(reused_input_tensors_[index]);

    input_tensor[0].name = InputInfoOfRuntime(0).name;

    if (!Infer(input_tensor, &reused_output_tensors_)) {
      FDERROR << "Failed to inference by runtime." << std::endl;
      return false;
    }

    (*results)[index].Clear();
    (*results)[index].Reserve(reused_output_tensors_[0].shape[0]);
    if (!postprocessor_.Run(reused_output_tensors_, &((*results)[index]))) {
      FDERROR << "Failed to postprocess the inference results by runtime."
              << std::endl;
      return false;
    }
  }
  return true;
}

} // namespace perception
} // namespace vision
} // namespace ultra_infer
